src/schema.js

```diff
@@ -138,52 +139,31 @@ class Schema {
   /**
    * https://github.com/frictionlessdata/tableschema-js#schema
    */
-  castRow(items, {failFast=false, skipConstraints=false}={}) {
-    // TODO: this method has to be rewritten
-    const headers = this.fieldNames
+  castRow(row, {failFast=false}={}) {
     const result = []
     const errors = []
 
-    if (headers.length !== items.length) {
-      const message = 'Row dimension doesn\'t match schema\'s fields dimension'
-      throw new TableSchemaError(message)
+    // Check row length
+    if (row.length !== this.fields.length) {
+      throw new TableSchemaError(
+        `The row with ${row.length} values does not match ` +
+        `the ${this.fields.length} fields in the schema`
+      )
     }
 
-    for (let i = 0, length = items.length; i < length; i += 1) {
+    // Cast row
+    for (const [index, [field, value]] of zip(this.fields, row).entries()) {
       try {
-        const field = this.getField(headers[i], i)
-        const value = field.castValue(items[i], {constraints: !skipConstraints})
-
-        // TODO: reimplement
-        // That's very wrong - on schema level uniqueness doesn't make sense
-        // and it's very bad to use it for exteral (by Table) monkeypatching
-        if (!skipConstraints) {
-          // unique constraints available only from Resource
-          if (this.uniqueness && this.uniqueHeaders) {
-            helpers.checkUnique(field.name, value, this.uniqueHeaders,
-                                     this.uniqueness)
-          }
-        }
-        result.push(value)
-      } catch (e) {
-        let error
-        switch (e.name) {
-          case 'UniqueConstraintsError':
-            error = new TableSchemaError(e.message)
-            break
-          default:
-            error = new TableSchemaError(
-              `Wrong type for header: ${headers[i]} and value: ${items[i]}`)
-        }
-        if (failFast === true) {
-          throw error
-        } else {
-          errors.push(error)
-        }
+        result.push(field.castValue(value))
+      } catch (error) {
+        error.columnNumber = index + 1
+        if (failFast) throw error
+        errors.push(error)
       }
     }
 
-    if (errors.length > 0) {
+    // Raise errors
+    if (errors.length) {
       const message = `There are ${errors.length} cast errors (see 'error.errors')`
```

Comment:
I wonder if this is better worded, `There are ${errors.length} type and format mismatch errors (see 'error.errors')`